core/camel-base/src/main/java/org/apache/camel/impl/converter/CoreTypeConverterRegistry.java

```diff
@@ -107,7 +107,7 @@ private <T> T fastConvertTo(Class<T> type, Exchange exchange, Object value) {
             return null;
         }
 
-        if (type.isInstance(value)) {
+        if (type.equals(value.getClass())) {
```

Comment:
This sounds really risky since `value` could be an instance of a subclass like in the next example:

Given:
```
    public interface Animal {
    }

    public static class Dog implements Animal {
    }
```
When:
```
        Dog pluto = new Dog();
        System.out.println("Is pluto an animal? Using isInstance " + (Animal.class.isInstance(pluto) ? "Yes" : "No"));
        System.out.println("Is pluto an animal? Using Equals " + (Animal.class.equals(pluto.getClass()) ? "Yes" : "No"));
```
Then
```
Is pluto an animal? Using isInstance Yes
Is pluto an animal? Using Equals No
```

Reply:
It's still gonna be checked for this later. This is just a duplicated one.

Reply:
More specifically: the code is doing this check twice. I'm just making the first check a bit lenient to avoid hitting a slow path in the JVM itself.